File: backend/app/schemas/spread.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

from app.schemas.common import BaseSchema


class NeighborFarmDetail(BaseSchema):
    """Anonymized or public detail of adjacent agricultural holding."""

    farm_id: str
    farm_name: str
    distance_km: float
    crop_type: str
    growth_stage: str
    latitude: float
    longitude: float
    is_downwind: bool
    active_threat_level: str = "LOW"


class NeighborsResponse(BaseSchema):
    """List of geographic neighbors within search radius."""

    target_farm_id: str
    target_farm_name: str
    search_radius_km: float
    total_neighbors: int
    neighbors: List[NeighborFarmDetail]


class SpreadRiskEdgeDetail(BaseSchema):
    """Directed graph edge representing transmission risk from an external source."""

    source_farm_id: str
    source_farm_name: str
    target_farm_id: str
    target_farm_name: str
    distance_km: float
    wind_alignment_factor: float = Field(description="-1.0 to 1.0 (1.0 = direct downwind corridor)")
    crop_similarity_score: float = Field(description="0.0 to 1.0 host compatibility")
    estimated_spread_risk: int = Field(ge=0, le=100, description="0-100 Potential Spread Risk Score")
    spread_risk_tier: str = Field(description="LOW, MEDIUM, HIGH, CRITICAL")
    source_pathogen: str
    estimated_arrival_days: Optional[int] = None
    explanation: str


class SpreadRiskGraphResponse(BaseSchema):
    """Directed Farm Risk Graph edges and contagion risk corridors."""

    farm_id: str
    farm_name: str
    overall_spread_threat_level: str
    max_estimated_spread_risk: int
    incoming_risk_edges: List[SpreadRiskEdgeDetail]
    wind_parameters: Dict[str, Any]
    disclaimer: str = Field(
        default="Estimated Spread Risk represents heuristic epidemiological dispersion modeling, not confirmed laboratory transmission.",
        description="Scientific transmission disclaimer",
    )


class RegionalHotspotDetail(BaseSchema):
    """Cluster zone with elevated regional disease or pest pressure."""

    hotspot_id: str
    name: str
    latitude: float
    longitude: float
    radius_km: float
    active_outbreaks_count: int
    dominant_pathogen: str
    hotspot_severity_level: str  # "MODERATE", "HIGH", "CRITICAL"
    affected_farms_count: int


class RegionalHotspotsResponse(BaseSchema):
    """Regional contagion cluster hotspots detected across the agricultural corridor."""

    total_hotspots: int
    evaluated_at: datetime
    hotspots: List[RegionalHotspotDetail]

File: frontend/src/types/spread.ts
export interface NeighborFarmDetail {
  farm_id: string;
  farm_name: string;
  distance_km: number;
  crop_type: string;
  growth_stage: string;
  latitude: number;
  longitude: number;
  is_downwind: boolean;
  active_threat_level: string;
}

export interface NeighborsResponse {
  target_farm_id: string;
  target_farm_name: string;
  search_radius_km: number;
  total_neighbors: number;
  neighbors: NeighborFarmDetail[];
}

export interface SpreadRiskEdgeDetail {
  source_farm_id: string;
  source_farm_name: string;
  target_farm_id: string;
  target_farm_name: string;
  distance_km: number;
  wind_alignment_factor: number;
  crop_similarity_score: number;
  estimated_spread_risk: number; // 0-100
  spread_risk_tier: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  source_pathogen: string;
  estimated_arrival_days?: number;
  explanation: string;
}

export interface SpreadRiskGraphResponse {
  farm_id: string;
  farm_name: string;
  overall_spread_threat_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  max_estimated_spread_risk: number;
  incoming_risk_edges: SpreadRiskEdgeDetail[];
  wind_parameters: {
    direction_degrees: number;
    speed_kmh: number;
    compass_heading: string;
  };
  disclaimer: string;
}

export interface RegionalHotspotDetail {
  hotspot_id: string;
  name: string;
  latitude: number;
  longitude: number;
  radius_km: number;
  active_outbreaks_count: number;
  dominant_pathogen: string;
  hotspot_severity_level: 'MODERATE' | 'HIGH' | 'CRITICAL';
  affected_farms_count: number;
}

export interface RegionalHotspotsResponse {
  total_hotspots: number;
  evaluated_at: string;
  hotspots: RegionalHotspotDetail[];
}
